engines/groovie/roq.cpp

```diff
@@ -124,31 +196,120 @@ uint16 ROQPlayer::loadInternal() {
-
-		for (int x = 0; x < _bg->w; x++) {
-			// Copy a pixel, checking the alpha channel first
-			if (_alpha && !(*in & 0xFF))
-				out++;
-			else if (_fg->h == 480 && *in == _vm->_pixelFormat.RGBToColor(255, 255, 255))
-				// Handle transparency in Gamepad videos
-				// TODO: For now, we detect these videos by checking for full screen
-				out++;
-			else
-				*out++ = *in;
+	// Calculate screen offset for normal / fullscreen videos and images
+	int screenOffset = 0;
+	if (_screen->h != 480) {
+		screenOffset = 80;
+	}
+
+	// Restore the background by data from the foreground. Only restore the area which was overwritten during the last frame
+	// Therefore we have the _restoreArea which reduces the area for restoring. We also use the _prevBuf to only overwrite the
+	// Pixels which have been written during the last frame. This means _restoreArea is just an optimization.
+	if (_alpha) {
+		if (!_restoreArea->isEmpty()) {
+			int width = _restoreArea->right - _restoreArea->left;
+			Graphics::Surface *screen = _vm->_system->lockScreen();
+			for (int line = _restoreArea->top; line < _restoreArea->bottom; line++) {
+				byte *dst = (byte *)screen->getBasePtr(_restoreArea->left, line + screenOffset);
+				byte *src = (byte *)_bg->getBasePtr(_restoreArea->left, line);
+				byte *prv = (byte *)_prevBuf->getBasePtr((_restoreArea->left - _origX) / _scaleX, (line - _origY) / _scaleY);
+				byte *ovr = (byte *)_overBuf->getBasePtr(_restoreArea->left, line);
+				for (int i = 0; i < width; i++) {
+					if (prv[kAIndex] != 0) {
+						copyPixel(dst, src);
+						copyPixelWithA(dst, ovr);
+					}
+					src += _bg->format.bytesPerPixel;
+					dst += _bg->format.bytesPerPixel;
+					prv += _bg->format.bytesPerPixel;
+					ovr += _bg->format.bytesPerPixel;
+				}
+			}
+			_vm->_system->unlockScreen();
+		}
+
+		// Reset _restoreArea for the next frame
+		_restoreArea->top = 480;
+		_restoreArea->left = 640;
+		_restoreArea->bottom = 0;
+		_restoreArea->right = 0;
+	}
+
+
+	// Select the destination buffer according to the given flags
+	int destOffset = 0;
+	Graphics::Surface *destBuf;
+	if (_flagOne) {
+		if (_flagTwo) {
+			destBuf = _overBuf;
+		} else {
+			destBuf = _bg;
+		}
+	} else {
+		destBuf = _vm->_system->lockScreen();
+		destOffset = screenOffset;
+	}
+
+	
+	// _origY and _origX may be negative (11th hour uses this in the chapel puzzle against Stauf)
+	int startX, startY, stopX, stopY;
+	calcStartStop(startX, stopX, _origX, _screen->w);
+	calcStartStop(startY, stopY, _origY, _screen->h);
+
+	for (int line = startY; line < stopY; line++) {
+		byte *in = (byte *)_currBuf->getBasePtr(MAX(0, -_origX) / _scaleX, (line - _origY) / _scaleY);
+		byte *inOvr = (byte *)_overBuf->getBasePtr(startX, line);
+		byte *out = (byte *)destBuf->getBasePtr(startX, line + destOffset);
+
+		for (int x = startX; x < stopX; x++) {
+			if (destBuf == _overBuf) {
+				copyPixelIfAlpha(out, in);
+			} else {
+				if (_interlacedVideo && (line % 2) && 0) {
```

Comment:
Put another fat `warning()` here, so it is not forgotten that this branch is disabled?

Reply:
I can just delete it, does anyone really wanna play with black lines instead of just filling them in?

Reply:
I think the videos look a little bit better with the black lines. But only if there is no scaling. Scaling the black lines looked very disruptive. As far as I could remember, I nearly tried every possible scaler during my tests.
I just added to code to be as much as close as possible to the original game.